### DISCUSSION TITLE
Disable TOTP

### DISCUSSION THREAD

**Discussion by catap** [dani-garcia/vaultwarden#4287] [Q&A]
Good day,

I m confused.

The basic idea of TOTP is make an authorisation with two factors, to reduce risk that the first factor (password) is compromised.

The feature to store TOTP inside password manager simple dimiss this idea. It turns Password + TOTP into two components password which is automatically filled by browser extention.

Is it possible to **disable** this feature for password stored in the server?

Thanks!

**Comment by BlackDex** [comment]
No, this is not possible without disabling all other premium features.

You can just choose to not use it. And store your 2FA/MFA tokens using other tools.

**Reply by catap** [reply]
Yes, in personal setup I can simple ignore this feature.

but in the case corporate setup, I can’t do it and such features will be widely used.

so, I’m looking for a way to break it ;)

**Reply by BlackDex** [reply]
I quickly tested it. But there is no other way then to disable Premium.
All other options seem to still keep the ability to enter a token, and then it will still show the result.

Without premium there also is no attachments for example.
So, i think it will be difficult to fully remove that feature.

Only hiding it in the web-vault will not work, since the other clients still show that field.

**Reply by catap** [reply]
@BlackDex is it possible to add feature that any OTP token will be replaced to random data when it's fetched from the server?

An idea to break that feature and prevent it from users to use it. Not working is good enough.

**Reply by BlackDex** [reply]
It's not something i would like to add to Vaultwarden as a feature.

You can try to manipulate the data in the database. By running a query over all ciphers which clears the specific totp field maybe.

**Reply by catap** [reply]
@BlackDex which makes things quite complicated.

My point: such feature destroy any value of TOTP, and shouldn't be available at all.

We can't change Bitwarden clients, but we can fix the server to make it possible to not use it.

**Reply by stefan0xC** [reply]
> We can't change Bitwarden clients, but we can fix the server to make it possible to not use it.

Indeed. You are free to compile the server with any modification you want.

**Reply by BlackDex** [reply]
To be fair, it's not a feature of Bitwarden to disable it in some way or another. And I still want to keep as close to Bitwarden as possible.

Also, i thought it was the whole idea to _destroy_ that value. I'm not sure what the difference in the end is between removing it via a sql script or not allowing it to be entered in the first place.

The problem with both is that the field to provide the string is still there. And that will not prevent users from trying to use it.

And, as @stefan0xC mentioned, you can always maintain a fork your self which has that feature.

**Reply by stefan0xC** [reply]
> No, this is not possible without disabling all other premium features.

Just took a look and it should be possible by simple setting this to `false`:
https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/db/models/cipher.rs#L211-L213

Question is if we want to make it configurable? (edit: Sorry, the referenced discussion was about hardening Vaultwarden itself.)

NOTE: This will not prevent users from saving the TOTP authenticator key in the field (that would be something Bitwarden would have to change). It will just not be as useful. (I mean you can also save the key in a different field so you can't prevent that.)

**Reply by stefan0xC** [reply]
To really prevent it, we'd have to filter out the `data` field that is sent by the clients when saving/updating a cipher and like @BlackDex said this is not really something that will be added to Vaultwarden.

**Comment by BlackDex** [comment]
> > No, this is not possible without disabling all other premium features.
> 
> Just took a look and it should be possible by simple setting this to `false`:
> 
> https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/db/models/cipher.rs#L211-L213
> 
> Question is if we want to make it configurable (as suggested in #4222)?
> 
> NOTE: This will not prevent users from saving the TOTP authenticator key in the field (that would be something Bitwarden would have to change). It will just not be as useful. (I mean you can also save the key in a different field so you can't prevent that.)

I tried to set that to false. And for me, i was still able to add a key and afterwards copy the token. I even set `useTotp` to false, which showed me i had to upgrade, but it still worked 😅

**Reply by stefan0xC** [reply]
> but it still worked 😅

Ah, I have not tried it. You are correct. Seems like this is also tied to whether the user has a premium account.
Cf. https://github.com/bitwarden/clients/blob/2511ae959a0a1979d2b9bd91d46d8940e845f8cf/libs/angular/src/vault/components/view.component.ts#L144

**Reply by catap** [reply]
so, it is impossible to disable, am I right?

**Reply by BlackDex** [reply]
It is.
The only way possible to prevent it from being used at all is by filtering the data field as @stefan0xC already mentioned.
And then it would be best to error out if someone used that field. It shouldn't break the input, but show the user a error.

But again, that is not something we will be adding to Vaultwarden.
I also think it has no link to #4222 at all. That is more a way to force secure 2FA's. Which then would make it more safe to allow storage of these tokens hehe ;).

**Reply by stefan0xC** [reply]
Yeah, well, it's impossible without also severely restricting the available premium options. But like we already said, if you want Vaultwarden to prevent saving anything in the field (`Login.Totp`) you'd have to write a filter, which would be technically possible but rather tedious (cf. the relevant [`update_cipher_from_data()`](https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/api/core/ciphers.rs#L351-L525) function) and not something we will support, or alternatively, you can write a script that clears the field in the database as suggested by @BlackDex (but since we just store the cipher data as `JSON` I would not really recommend that approach).

**Reply by catap** [reply]
@stefan0xC I don't want to prevent saving. The best outcome is make it impossible to use. I not sure if client read token which was sent, and not simple use it to compute the first TOTP to confirm that was installed.

If client do not re-fetch token from server to confirm that it was stored correctly, well, when idea to replace it to garbage on server makes things quite ugly. User may add TOTP but at some point it won't work.

My goal is make impossible to users to share passwords with TOTP, because they will do it. Because it is quite comfortable to use ;)

**Reply by stefan0xC** [reply]
> And then it would be best to error out if someone used that field. It shouldn't break the input, but show the user a error.

I don't think the Bitwarden clients are capable of doing that. (And if we had that level of control over all the clients we could just prevent users from showing/filling out the field in the first place.) :shrug:

**Reply by stefan0xC** [reply]
@catap Like I said you can compile Vaultwarden yourself with (I think) setting both of these values to `false`: https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/db/models/user.rs#L248 and https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/db/models/cipher.rs#L213

This _should_ disable the generation of TOTP in the Bitwarden clients and thus incentivize users not to save the key in the first place (which they would still be able to do).

**Reply by BlackDex** [reply]
You also need to set:

https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/db/models/organization.rs#L187

and

https://github.com/dani-garcia/vaultwarden/blob/ad1d65bdf8a15d3bc1eac076ab4ab394777204ee/src/db/models/organization.rs#L354

to false

Else the user still gets premium.
But, that will also disable attachment support for personal vaults.
Which might not be an issue if you want to only force users to use the organization.

**Reply by BlackDex** [reply]
It is maybe even possible to fetch this using a reverse proxy and change the values there.

**Reply by stefan0xC** [reply]
> Else the user still gets premium.

I have not tested it but I'm not sure that's the case. :thinking:  I mean we have hard coded the `Premium` option to `true` for all users in Vaultwarden, so if this is changed to `false` it shouldn't make a difference whether users of an organization will get premium or not (but it shouldn't hurt to set those to `false` too).

**Reply by BlackDex** [reply]
I tested it, and users get premium from the org, amd thus are still able to use it.

**Reply by catap** [reply]
Thanks, it's clear that it is impossible to disable this feature.

BTW I quite surprised that this exists at all, let me be honest.
